src/OrchardCore/Orchard.Environment.Navigation/NavigationBuilder.cs

```diff
@@ -55,9 +55,27 @@ public NavigationBuilder Remove(MenuItem item)
             return this;
         }
 
-        public List<MenuItem> Build()
+        public NavigationBuilder Remove(Predicate<MenuItem> match)
+        {
+            RemoveRecursive(Contained, match);
+            return this;
+        }
+
+        public virtual List<MenuItem> Build()
         {
             return (Contained ?? new List<MenuItem>()).ToList();
         }
+
+        private static void RemoveRecursive(List<MenuItem> menuItems, Predicate<MenuItem> match)
+        {
+            menuItems.RemoveAll(match);
+            foreach (var menuItem in menuItems)
```

Comment:
menuItems.RemoveAll(match);
foreach (var menuItem in menuItems){
...
No, I'm updating it before iteration